packages/core/src/factory.ts

```diff
@@ -18,62 +22,71 @@ export function factory<Resource, Input, Output, Config extends IPrismConfig>(
-                  map(sec => [sec]),
-                  getOrElse<IPrismDiagnostic[]>(() => [])
-                )
+        TaskEither.chain(resource =>
+          TaskEither.fromEither(
+            pipe(
+              sequenceValidation(
+                config.validateRequest ? components.validateInput({ resource, element: input }) : Either.right(input),
+                config.checkSecurity ? validateSecurity(input, resource) : Either.right(input)
+              ),
+              Either.fold(
+                validations =>
+                  Either.right<Error, { resource: Resource; validations: IPrismDiagnostic[] }>({
+                    resource,
+                    validations,
+                  }),
+                () =>
+                  Either.right<Error, { resource: Resource; validations: IPrismDiagnostic[] }>({
+                    resource,
+                    validations: [],
+                  })
               )
-            : inputValidations;
-
-          const outputLocator = isProxyConfig(config)
+            )
+          )
+        ),
+        TaskEither.chain(({ resource, validations }) => {
+          const produceOutput = isProxyConfig(config)
             ? components.forward(input, config.upstream.href)
             : TaskEither.fromEither(
                 components.mock({
                   resource,
                   input: {
-                    validations: inputValidations,
+                    validations,
                     data: input,
                   },
                   config: config.mock,
                 })(components.logger.child({ name: 'NEGOTIATOR' }))
               );
 
           return pipe(
-            outputLocator,
-            TaskEither.map(output => ({ output, resource }))
+            produceOutput,
+            TaskEither.map(output => ({ output, resource, validations }))
           );
         }),
-        TaskEither.map(({ output, resource }) => {
-          let outputValidations: IPrismDiagnostic[] = [];
-          if (config.validateResponse) {
-            outputValidations = components.validateOutput({
-              resource,
-              element: output,
-            });
-          }
+        TaskEither.map(({ output, resource, validations }) => {
+          const outputValidations = pipe(
+            config.validateResponse,
+            Option.fromPredicate(t => t),
+            Option.chain(() =>
+              Option.fromEither(
+                pipe(
+                  components.validateOutput({ resource, element: output }),
+                  Either.swap
+                )
+              )
+            ),
+            Option.map<NonEmptyArray<IPrismDiagnostic>, IPrismDiagnostic[]>(t => t),
+            Option.getOrElse<IPrismDiagnostic[]>(() => [])
+          );
 
           return {
             input,
             output,
             validations: {
-              input: inputValidations,
+              input: validations,
```

Comment:
maybe be frank about `validations` from the beginning and call it `inputValidations`? so here, we would end up with:
```
input: inputValidations,
output: outputValidations,
```

Reply:
Good point. 277d9b7